### DISCUSSION TITLE
Mapping incomplete for "Company" lookup

### DISCUSSION THREAD

**Discussion by ollitheorca** [microsoft/Dual-write-automations#20] [Q&A]
Hi,

I have observed a recurring issue when running through my config file for all my maps. There seems to be an issue with the Company field not being properly refreshed/populated when a DW map is being executed for the target state. And that causes a spiral of errors on related DW maps going forward.

It doesn't matter if I choose Deploy/Initial sync, or Start maps as target state. I still see the same issue.

Let me try to give you an example of my issue when going for a simple "start maps" state:

I have the following custom map that is not being properly started and instead showing this error:
![image](https://github.com/microsoft/Dual-write-automations/assets/129459105/e1c21052-813f-4c1c-bfc9-699bbc2788c3)

When opening the map I get this field map error:
![image](https://github.com/microsoft/Dual-write-automations/assets/129459105/3590205b-84ed-4db9-ae14-d2b8f1a96f20)

So, essentially DW is complaining about the related DW map not being configured correctly (Fee types - export to Customer Engagement (CE) (flx_feetypes) even though it actually got through to running state. 

If we open the Fee types - export to Customer Engagement (CE) (flx_feetypes) map we can see the following:
![image](https://github.com/microsoft/Dual-write-automations/assets/129459105/506672ed-1d2e-4a4b-9b66-aa20ca09cffc)

The company field map is incomplete. However, DW handles this field map automatically and it doesn't need to be added to the DW field mapping list (at least AFAIK).

Clicking "Refresh tables" while it is still in running mode will actually clear the company field map error. And now going back to the initial DW map, Fee attributes (flx_feebasedatatables) and clicking refresh will now clear all the mapping errors.

Now, this is obviously annoying to deal with as I understood the tool should properly refresh the table maps. Yet it is pretty quick and simple to resolve. 

However, I have other table maps with the exact same issue only that these have way more relationships to other maps, resulting in a spiral of field mapping issues. Take this one as an example:
![image](https://github.com/microsoft/Dual-write-automations/assets/129459105/d1b6e1a4-58a9-470f-b68d-8b594bc5117c)

Working back through the affected (already running) maps will always show that the highest "parent" map always has the company field as the only incomplete mapping. Refreshing the table will fix that, and now you can start working your way back down the spiral refreshing each of them, which will clear the incomplete mappings and in the end result in the initial map (that was not able to be run) to be cleared for any of the errors.

I don't understand why some maps are able to be set into running mode whilst others do not. For maps with a lot of related tables, this is very frustrating and time-consuming to resolve. :-)

Please let me know if I can elaborate or share anything to give some insight into why I get this issue.

Any help or assistance would be very appreciated. 👍

**Comment by Hax1337** [comment]
@ollitheorca thanks for reaching out. 

Is the company Id field in the key list in your configuration file?  The way how the helper sets the keys is that it sets both fields at the same time wiping out the current setup.

This potentially leads to your error youre having if the company ID field is not added into the Key list in your config. 

2 things you can try:

1. Add the company field to the mapping
2. Remove all keys in the mapping, this will not apply keys and then uses refresh table to assign them

**Reply by ollitheorca** [reply]
Hi @Hax1337, 

Thanks for the swift response.

Yes, the company field is included in the key in the config file.

Here is the key for the "feebasedatatables - Fee attributes": (the one failing to go into running state)

<Map master="FO" mapName="flx_feebasedatatables - Fee attributes"
    version="2023.4.0.5" group="All" authors="FlexProperty" keys="flx_company.cdm_companycode,flx_feetypeid.flx_feetypename,flx_feetypeid.flx_company.cdm_companycode" />

Here is the key for the related Fee Type DW map (which is running):

<Map master="FO" mapName="flx_feetypes - Fee types - export to Customer Engagement (CE)"
		version="1.0.0.0" group="All" authors="DefaultPublisheryavicaflexdev"
		keys="flx_feetypename, flx_company" />

So essentially, **the key** for "feebasedatatables - Fee attributes" is the Lookup field to **Fee Type + Company**.

But since I'm seeing this error across all my maps, the only difference is that some maps are able to go through to running state (even tho they have the company error), while others are not, you then suggest I remove the keys for all my maps and let the "refresh" determine the keys instead? Is that correct understood? :-)

**Reply by Hax1337** [reply]
Thats interesting I would like to definately get to the bottom of this. 

few questions: 
When you do the export, and compare the map to your current configuration, do you see any differences? 

And yes, if you remove the keys it will not apply them but still refreshes table.

Are you able to provide a log for this particular map?  <ou can also finde me on LinkedIn if you do not want to share it here. 
https://www.linkedin.com/in/max-hentschel-2b37b858/

**Reply by ollitheorca** [reply]
Hello Max,

My apologies for the delay in responding to your message.

Upon further testing, I was able to identify the reason behind the issue I had mentioned earlier. It seems that on our development server, some of our individual DW maps had different publishers. As a result, when we exported the configuration file, it showed the different publishers in the file.

However, all our maps now go into one solution, and when we deploy that solution to another environment, all the DW maps get the publisher from the solution, even if some of the individual maps had different publishers on the dev server.

To fix the issue, we changed the publisher in the config file to match the solution publisher. :-)

**Reply by Hax1337** [reply]
Amazing, thanks for the update, I can check if there can be a message added when a certain publisher with the version is not found. That should make it easiert to determine then. 

Ill close this
